### DISCUSSION TITLE
RFC: Add ngrxOnResolveEffects lifecycle method to Effects

### DISCUSSION THREAD

**Discussion by hizliemre** [ngrx/platform#3390] [Ideas]
This lifecycle hook will be running like onInitEffects, but onResolvedEffects hook will be running on every resolving phase when using EffectSources addEffects method. 

use case example : https://stackblitz.com/edit/angular-ivy-tentrb

This is the implementation;
```typescript
// lifecycle_hooks.ts
export declare interface OnResolveEffects {
  /**
   * @description
   * Action to be dispatched after the effect is registered.
   */
  ngrxOnResolveEffects(): Action;
}

export const onResolveEffects: keyof OnResolveEffects = 'ngrxOnResolveEffects';

export function inOnResolveEffects(instance: any): instance is OnResolveEffects {
  return isFunction(instance, onResolveEffects);
}
```
```typescript
// effect_sources.ts end of toActions method.
const init$ = source$.pipe(take(1),
  filter(isOnInitEffects),
  map((instance) => instance.ngrxOnInitEffects())
);

const resolve$ = source$.pipe(
  filter(inOnResolveEffects),
  map((instance) => instance.ngrxOnResolveEffects())
);

return merge(effect$, init$, resolve$);

```

**Comment by timdeschryver** [comment]
What's the use-case for introducing the `ngrxOnResolveEffects` hook?

**Reply by hizliemre** [reply]
> What's the use-case for introducing the `ngrxOnResolveEffects` hook?

You can check this example. https://stackblitz.com/edit/angular-ivy-tentrb

I will replace ngrxOnInitEffects with the new hook.

**Reply by hizliemre** [reply]
I handled this in this pull request.
https://github.com/ngrx/platform/pull/3391

**Reply by timdeschryver** [reply]
Yea that's a good demo, but why do you need this to be a part of the API?

**Reply by hizliemre** [reply]
@timdeschryver  check my stackblitz demo again and look to page 2 state.

Because I don't want to share my private fields in effect instance - think about adding effects conditionally -, so I cant dispatch my initialize action in my component OnInit method. I have to be a part of API. When I saw OnInitEffects I was very happy but it's not worked as I expected. 

I think my hook can be very useful for many things. This will give us flexibility.

**Reply by hizliemre** [reply]
And I know I can add one more effect for this but it's not looking good to readability and maintainability. To do this a lifecycle hook would be very nice.

- BeforeInitialize 
- Initialize
- AfterInitialize

this smells bad. If I try to use my private field for many effects I don't want create double action for handle my problem. Its horrible.

**Comment by hizliemre** [comment]
What are your thoughts @brandonroberts ?

**Comment by nevzatopcu** [comment]
Any update on this @brandonroberts ? cc: @timdeschryver 

In such case we need to have a lifecycle hook to dispatch an action that has some private props. Obviously, we can solve that issue by dispatching the action on component but alas it needs to have the props to dispatch the action.

**Comment by hizliemre** [comment]
If you want, I am ready to update my pull request. Also, I've updated the stackblitz example to explain more clearly.

Could you check it again, please? @timdeschryver

**Reply by timdeschryver** [reply]
I don't see a big reason to add this.

- the workaround is simple and offers more flexibility
- the Angular API (and also NgRx) is moving to more a functional way, which also means that there are no lifecycles (https://github.com/ngrx/platform/pull/3669)
- to prevent duplication you could wrap the `addEffect` and `init` methods in a method, or within an `inject`
